### PR TITLE
Add scope pruning

### PR DESCRIPTION
**Pull request by defektive**
I am sure I missed something... 

Moved scope related things to own package
Added tests <3

* * * *

Test Plan
---------

Test `arsenic discover` and `arsenic analyze`

### CODE REVIEW

File: lib/scope/scope.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+}
+
+var asScope = &scope{}
+var asScopeInit = false
+
+func getScope() *scope {
+	return asScope
+}
+
+func (s *scope) loadExplicitDomains () {
```
Comment (NoF0rte) [inline lib/scope/scope.go]: You should run `gofmt`

File: lib/scope/scope.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	blacklistedRootDomains []string
+	blacklistedDomains []string
+	explicitDomainsLoaded bool
+
+	hostIPsExplicitlyInScope []string
+	hostIPsExplicitlyInScopeLoaded bool
+}
+
+var asScope = &scope{}
+var asScopeInit = false
```
Comment (NoF0rte) [inline lib/scope/scope.go]: Nit: This just for the test? `go-staticcheck` tells me that it isn't being used anywhere

File: lib/scope/scope.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+		util.ReadLineByLine("scope-domains.txt", func(line string) {
+			line = normalizeScope(line, "domain")
+			rootDomain, _ := publicsuffix.EffectiveTLDPlusOne(line)
+
+			domainsMap[line] = true
+			rootDomainsMap[rootDomain] = true
+		})
+
+		for domain, _ := range domainsMap {
```
Comment (NoF0rte) [inline lib/scope/scope.go]: Nit: Can remove `, _` since you aren't using the bool value

File: lib/scope/scope.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+			s.domainsExplicitlyInScope = append(s.domainsExplicitlyInScope, domain)
+		}
+
+		for rootDomain, _ := range rootDomainsMap {
```
Comment (NoF0rte) [inline lib/scope/scope.go]: Nit: Same as above

File: lib/scope/scope.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+var asScope = &scope{}
+var asScopeInit = false
+
+func getScope() *scope {
+	return asScope
+}
+
+func (s *scope) loadExplicitDomains () {
+	if !s.explicitDomainsLoaded {
+		domainsMap := map[string]bool{}
```
Comment (NoF0rte) [inline lib/scope/scope.go]: Nit: Could use `Set` if you wanted to

File: lib/scope/scope.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+func (s *scope) loadExplicitScopeIPs () {
+	if !s.hostIPsExplicitlyInScopeLoaded {
+		hostIPMap := map[string]bool{}
+
+		util.ReadLineByLine("scope-ips.txt", func(line string) {
+			line = normalizeScope(line, "ip")
+			hostIPMap[line] = true
+		})
+
+		for hostIP, _ := range hostIPMap {
```
Comment (NoF0rte) [inline lib/scope/scope.go]: Nit: Same as above

File: lib/scope/scope.go [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		}
+	}
+
+	return false
+}
+
+func (s *scope) IsBlacklistedDomain (checkDomain string) bool {
+	s.loadExplicitDomains()
+	for _, badDomain := range s.blacklistedDomains {
+		if strings.EqualFold(badDomain, checkDomain) {
```
Comment (NoF0rte) [inline lib/scope/scope.go]: Don't want to use regex like we were doing before?

**Review by NoF0rte — CHANGES_REQUESTED** [review]
Review my comments

File: lib/scope/scope.go (hunk 1 above)
Comment (defektive) [reply lib/scope/scope.go:28]: no changes...

File: lib/scope/scope.go (hunk 2 above)
Comment (defektive) [reply lib/scope/scope.go]: i think this can be deleted. It was remnants from earlier code

**Review by NoF0rte — APPROVED** [review]

**Pull request merged**